src/librustdoc/html/markdown.rs

```diff
@@ -353,145 +212,300 @@ pub fn render(w: &mut fmt::Formatter,
-            let text = str::from_utf8(text).unwrap();
-            let lines = text.lines().map(|l| {
-                stripped_filtered_line(l).unwrap_or(l)
-            });
-            let text = lines.collect::<Vec<&str>>().join("\n");
-            let line = tests.get_line() + line;
-            let filename = tests.get_filename();
-            tests.add_test(text.to_owned(),
-                           block_info.should_panic, block_info.no_run,
-                           block_info.ignore, block_info.test_harness,
-                           block_info.compile_fail, block_info.error_codes,
-                           line, filename);
+                break
+            }
+        }
+        if content.is_empty() {
+            return
         }
+        buffer.push_str(&format!("<thead><tr>{}</tr></thead>", content.replace("td>", "th>")));
     }
 
-    extern fn header(_ob: *mut hoedown_buffer,
-                     text: *const hoedown_buffer,
-                     level: libc::c_int, data: *const hoedown_renderer_data,
-                     _: libc::size_t) {
-        unsafe {
-            let opaque = (*data).opaque as *mut hoedown_html_renderer_state;
-            let tests = &mut *((*opaque).opaque as *mut ::test::Collector);
-            if text.is_null() {
-                tests.register_header("", level as u32);
+    fn table(parser: &mut Parser, buffer: &mut String, toc_builder: &mut Option<TocBuilder>,
+             shorter: bool) {
+        let mut content = String::new();
+        let mut rows = String::new();
+        loop {
+            let event = parser.next();
+            if let Some(event) = event {
+                match event {
+                    Event::End(Tag::Table(_)) => break,
+                    Event::Start(Tag::TableHead) => {
+                        head(parser, &mut content, toc_builder, shorter);
+                    }
+                    Event::Start(Tag::TableRow) => {
+                        row(parser, &mut rows, toc_builder, shorter);
+                    }
+                    _ => {}
+                }
             } else {
-                let text = (*text).as_bytes();
-                let text = str::from_utf8(text).unwrap();
-                tests.register_header(text, level as u32);
+                break
             }
         }
+        buffer.push_str(&format!("<table>{}{}</table>",
+                                 content,
+                                 if shorter || rows.is_empty() {
+                                     String::new()
+                                 } else {
+                                     format!("<tbody>{}</tbody>", rows)
+                                 }));
     }
 
+    fn looper<'a>(parser: &'a mut Parser, buffer: &mut String, next_event: Option<Event<'a>>,
+                  toc_builder: &mut Option<TocBuilder>, shorter: bool) -> bool {
+        if let Some(event) = next_event {
```

Comment:
Doesn't change much from my point of view... I think this syntax is actually better.